### PR TITLE
[Merged by Bors] - Implement Debug for Res and ResMut

### PR DESCRIPTION
**Pull request by forbjok**
This commit adds blanket implementations of Debug for Res and ResMut, as discussed in https://github.com/bevyengine/bevy/issues/2048.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
As I mentioned in #2048, could you quickly add this for NonSend, NonSendMut and Local as well? Should be identical as I understand it.

**Comment by forbjok** [comment]
Done. There was already an implementation for NonSendMut, but I updated it to show the wrapper, like the others.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/system/system_param.rs [hunk 1]
```diff
@@ -174,6 +175,19 @@ pub struct Res<'w, T: Component> {
...
+impl<'w, T: Component> Debug for Res<'w, T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("Res(")?;
+        Debug::fmt(self.value, f)?;
+        f.write_str(")")?;
+
+        Ok(())
```
Comment (NathanSWard) [inline crates/bevy_ecs/src/system/system_param.rs:187]: These (and all other `Debug` impls) can all be reduce to a single line:
```rust
f.write_fmt(format_args!("Res({:?})", self.value))
```

Comment (NathanSWard) [reply crates/bevy_ecs/src/system/system_param.rs:187]: Also, when you `#[derive(Debug)]` for a struct, normally the format is: `StructName { <values> }`

So perhaps we should stick with this style, unless we want to explicitly mimic a tuple-like formatting.

Tuple-like:
```rust
"Res(<value>)"
```

Struct-like:
```rust
"Res { value: <value> }"
```

Comment (forbjok) [reply crates/bevy_ecs/src/system/system_param.rs]: Changed all the implementations to use this method.

Comment (mockersf) [reply crates/bevy_ecs/src/system/system_param.rs]: rust debug actually has helper for that https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_tuple for tuple
```
        fmt.debug_tuple("Res")
            .field(&self.value)
            .finish()
```
or https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct for struct

they let caller control more of the formatting, like precision, pretty print, ...

File: crates/bevy_ecs/src/system/system_param.rs [hunk 2]
```diff
@@ -174,6 +175,15 @@ pub struct Res<'w, T: Component> {
     change_tick: u32,
 }
 
+impl<'w, T: Component> Debug for Res<'w, T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("Res({:?})", self.value))
```
Comment (mockersf) [inline crates/bevy_ecs/src/system/system_param.rs:183]: rust debug actually has helper for that https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_tuple for tuple
```
        fmt.debug_tuple("Res")
            .field(&self.value)
            .finish()
```
or https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct for struct

they let caller control more of the formatting, like precision, pretty print, ...

**Review by mockersf — CHANGES_REQUESTED** [review]

**Review by mockersf — APPROVED** [review]

**Pull request closed without merging**